pkg/resource/v1/certconfig/update_test.go

```diff
@@ -1 +1,195 @@
+				newCertConfigWithVersion("cluster-1", certs.APICert, "1.0.0"),
+				newCertConfigWithVersion("cluster-1", certs.CalicoCert, "1.0.0"),
+				newCertConfigWithVersion("cluster-1", certs.EtcdCert, "1.0.0"),
+			},
+			desiredState: []*v1alpha1.CertConfig{
+				newCertConfigWithVersion("cluster-1", certs.APICert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.CalicoCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.EtcdCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.FlanneldCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.NodeOperatorCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.PrometheusCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.ServiceAccountCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.WorkerCert, "1.2.0"),
+			},
+			expectedCertConfigs: []*v1alpha1.CertConfig{
+				newCertConfigWithVersion("cluster-1", certs.APICert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.CalicoCert, "1.2.0"),
+				newCertConfigWithVersion("cluster-1", certs.EtcdCert, "1.2.0"),
+			},
+			errorMatcher: nil,
+		},
+		{
+			name: "case 4: desiredState is wrong type",
+			clusterGuestConfig: &v1alpha1.ClusterGuestConfig{
+				ID: "cluster-1",
+			},
+			currentState: []*v1alpha1.CertConfig{
+				newCertConfig("cluster-1", certs.APICert),
+				newCertConfig("cluster-1", certs.CalicoCert),
+				newCertConfig("cluster-1", certs.EtcdCert),
+			},
+			desiredState: []string{
+				"foo",
+				"bar",
+				"baz",
+			},
+			expectedCertConfigs: []*v1alpha1.CertConfig{},
+			errorMatcher:        IsWrongType,
+		},
+		{
+			name: "case 5: currentState is wrong type",
+			clusterGuestConfig: &v1alpha1.ClusterGuestConfig{
+				ID: "cluster-1",
+			},
+			currentState: []string{
+				"foo",
+				"bar",
+				"baz",
+			},
+			desiredState: []*v1alpha1.CertConfig{
+				newCertConfig("cluster-1", certs.APICert),
+				newCertConfig("cluster-1", certs.CalicoCert),
+				newCertConfig("cluster-1", certs.EtcdCert),
+			},
+			expectedCertConfigs: []*v1alpha1.CertConfig{},
+			errorMatcher:        IsWrongType,
+		},
+	}
+
+	logger, err := micrologger.New(micrologger.DefaultConfig())
+	if err != nil {
+		t.Fatalf("micrologger.New() failed: %#v", err)
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			r, err := New(Config{
+				G8sClient:   fake.NewSimpleClientset(),
+				K8sClient:   clientgofake.NewSimpleClientset(),
+				Logger:      logger,
+				ProjectName: "cluster-operator",
+				ToClusterGuestConfigFunc: func(v interface{}) (*v1alpha1.ClusterGuestConfig, error) {
+					return v.(*v1alpha1.ClusterGuestConfig), nil
+				},
+			})
+
+			if err != nil {
+				t.Fatalf("Resource construction failed: %#v", err)
+			}
+
+			certConfigs, err := r.newUpdateChange(context.TODO(), tt.clusterGuestConfig, tt.currentState, tt.desiredState)
+
+			switch {
+			case err == nil && tt.errorMatcher == nil: // correct; carry on
```

Comment:
I like this switch case a lot :+1:
Glad we did the discussion in fmt.

Reply:
Yes, I was always reinventing the way I handle errorMatcher :D